lib/uklock/mutex.c

```diff
@@ -1,8 +1,69 @@
 #include <uk/mutex.h>
 
+#ifdef CONFIG_LIBUKLOCK_MUTEX_METRICS
+#include <uk/init.h>
+#include <uk/store.h>
+#include <uk/print.h>
+#include <stdlib.h>
+
+struct uk_mutex_metrics mutex_metrics;
+spinlock_t              mutex_metrics_lock;
+#endif /* CONFIG_LIBUKLOCK_MUTEX_METRICS */
+
 void uk_mutex_init(struct uk_mutex *m)
 {
 	m->lock_count = 0;
 	m->owner = NULL;
 	uk_waitq_init(&m->wait);
+
+#ifdef CONFIG_LIBUKLOCK_MUTEX_METRICS
+	ukarch_spin_lock(&mutex_metrics_lock);
+	mutex_metrics.active_unlocked++;
+	ukarch_spin_unlock(&mutex_metrics_lock);
+#endif /* CONFIG_LIBUKLOCK_MUTEX_METRICS */
+}
+
+#ifdef CONFIG_LIBUKLOCK_MUTEX_METRICS
+/**
+ * Retrieves mutex-related metrics.
+ * @param cookie Unused.
+ * @param dst Reference to metrics copy buffer (allocated here).
+ * @return 0 on success, !0 otherwise.
+ *
+ * NOTE: this function is registered as a ukstore static store's getter
+ */
+static int uk_mutex_metrics_get(void *cookie __unused, __uptr *dst)
+{
+	/* allocate memory for copy of stats */
+	struct uk_mutex_metrics *to_send =
+		malloc(sizeof(struct uk_mutex_metrics));
+	if (!to_send) {
+		uk_pr_err("Unable to allocate memory for metrics structure\n");
+		return 1;
+	}
+
+	/* create copy */
+	memmove(dst, &mutex_metrics, sizeof(mutex_metrics));
+	*dst = (__uptr) to_send;
+
+	return 0;
 }
+UK_STORE_STATIC_ENTRY(mutex_stats, uptr, uk_mutex_metrics_get, NULL);
```

Comment:
Also, remove this, as it is related to `ukstore`. It can be introduced back as a separate `ukstore` PR in the future.